colossalai/pipeline/schedule/generate.py

```diff
@@ -0,0 +1,128 @@
+        self.microbatch_offset: Optional[int] = None
+        self.num_microbatches: Optional[int] = None
+
+    def load_batch(self, data_iter: Iterable, device: Optional[torch.device] = None) -> None:
+        """Load a batch from data iterator.
+
+        Args:
+            data_iter (Iterable): Data iterator.
+            device (Optional[torch.device], optional): Target device. Defaults to None.
+        """
+        batch = next(data_iter)
+        if device is not None:
+            batch = tree_map(partial(to_device, device=device), batch)
+        self.batch = batch
+        self.batch_size = get_batch_size(batch)
+        self.microbatch_offset = 0
+        assert self.batch_size % self.microbatch_size == 0, \
+            f"Batch size should divided by the number of microbatches, {self.batch_size}, {self.num_microbatches}"
+        self.num_microbatches = self.batch_size // self.microbatch_size
+        self.round = self.num_microbatches // self.stage_manager.num_stages
+
+    def load_micro_batch(self) -> Any:
+        """Load a micro batch from the current batch.
+
+        Returns:
+            Any: Micro batch.
+        """
+        micro_batch = get_micro_batch(self.batch, self.microbatch_offset, self.microbatch_size)
+        self.microbatch_offset += self.microbatch_size
+        return tree_map(partial(to_device, device=get_current_device()), micro_batch)
+
+    def _prepare_stage_inputs(self):
+        # first stage and in prefill phase
+        if self.stage_manager.is_first_stage() and self.mb_manager.cur_state is PREFILL:
+            pre_stage_out = None
+            model_inputs = self.load_micro_batch()
+            hidden_states = None
+        # first stage and in generate phase
+        elif self.stage_manager.is_first_stage():
+            pre_stage_out = self.comm.recv_forward()
+            model_inputs = self._prepare_next_token(pre_stage_out)
+            hidden_states = None
+        # not first stage and in gererate phase
+        else:
+            pre_stage_out = self.comm.recv_forward()
+            model_inputs = {
+                'past_key_values': self.mb_manager.cur_kv_cache
+            } if self.mb_manager.cur_kv_cache is not None else None
+            hidden_states = pre_stage_out
+        return pre_stage_out, model_inputs, hidden_states
+
+    def _prepare_next_token(self, input_ids):
+        new_mask = self.mb_manager.cur_descrption.attn_mask
+        new_mask = torch.cat((new_mask, torch.ones((new_mask.shape[0], 1), dtype=torch.int64).cuda()), dim=-1)
+        self.mb_manager.cur_descrption.attn_mask = new_mask
+        past_key_values = self.mb_manager.cur_descrption.kv_cache
+
+        return dict(input_ids=input_ids['new_token'], attention_mask=new_mask, past_key_values=past_key_values)
+
+    def get_token_id(self, hidden_state: torch.Tensor) -> torch.Tensor:
+        last_hidden_state = hidden_state[:, -1]
+        input_ids = torch.argmax(last_hidden_state, dim=-1).unsqueeze(1)
```

Comment:
It's greedy search, right? We have to support various generation args of huggingface/transformers. Greedy search is used only when `do_sample=False` and `num_beams=1`. Please add such kwargs to engine.

Reply:
You could first add these arguments and implement them in the next PR.

Reply:
done